### DISCUSSION TITLE
is there a way to update BrowserAuthOptions() dynamically?

### DISCUSSION THREAD

**Discussion by KeeganDC** [milestonetg/msal-js-dart#40] [Q&A] [answered]
`PublicClientApplication` needs to be instantiated before `MaterialApp` when using the redirect flow
but I would like to instantiate it dynamically via API

Just wondering if there is a way to update `BrowserAuthOptions()` or `clientId` after `MaterialApp` is created?

**Comment by Francessco121** [comment, marked as answer]
Unfortunately no. MSAL doesn't provide a way to update configuration after a `PublicClientApplication` is created, and you need to call `PublicClientApplication.handleRedirectFuture` before the `MaterialApp` router nukes the browser URL hash, which is required by MSAL to complete a redirect login.

The best workaround that I'm aware of is to save the current value of `window.location.hash` before `runApp` and pass it into your Flutter application. Then, you can instantiate MSAL whenever you want and pass that saved hash string to `handleRedirectFuture` manually (it takes an optional argument for the hash). Also as an aside, msal.js has had a v3 in the works for a bit, and I just checked to make sure that `handleRedirectFuture` does indeed still have this optional argument there, so hopefully no need to worry about them removing this in the future.

**Note:** If you do the above workaround, it's probably worth making sure that you don't store that URL hash after MSAL has been initialized, since it does contain an authentication token after a successful login. I'm not aware of how someone could exploit that being in memory, but I thought it was worth mentioning.

One last thing: `loginPopup` is unaffected by this issue and will work regardless, if you can get away with only using the popup flows.

Hope this helps!
